lsp-fiber/src/rpc.ml

```diff
@@ -190,19 +184,31 @@ struct
         Fiber.return
           (Jsonrpc_fiber.Reply.now (Jsonrpc.Response.error req.id error), state)
       | Ok (In_request.E r) ->
+        let cancel = Fiber.Cancel.create () in
         let+ response, state =
-          let cancel = Fiber.Cancel.create () in
-          Fiber.finalize
-            (fun () ->
-              Fiber.Var.set cancel_token cancel (fun () ->
-                  Table.replace t.pending req.id cancel;
-                  h_on_request.on_request t r))
-            ~finally:(fun () ->
-              Table.remove t.pending req.id;
```

Comment:
I think this cleanup is still needed if `h.on_request.on_request` raises instead of producing a value `Now` or `Later`.

We need to wrap this with a `Fiber.map_reduce_errors` and clean up `t.pending` in case of an error.

Reply:
I didn't use the `map_reduce_errors` function because it returns a `result` type and I don't know what that function should return in case of an `Error`. Perhaps the proper thing to do is for this function to also return a `result` but then we would have to handle the `Error` case somewhere else and that would require more changes.

Instead, I used the `map_and_reraise` function to clean up the `t.pending` and just re-raise the exception.

Does that make sense?

Reply:
Ah interesting, I hadn't thought about that but it definitely makes sense. Thanks :)